### ISSUE TITLE
[Java] C2D messages in the queue processing issues

### ISSUE THREAD

**Issue by dluc** [Azure/azure-iot-sdks#994]
When a service sends a number of messages to a device that is not connected, the messages are enqueued. When the device connects:

* if the service is still sending messages, these are received before the messages in the queue
* not all the messages in the queue are processed
* messages in the queue that happen to be processed are out of order
* messages are dequeued very slowly

Code: https://gist.github.com/dluc/974eae8bbfac9cb5682c9447e487bce5

Example output.

```
[00:01:45.793] Creating sender client
[00:01:46.005] Open sender connection
[00:01:46.009] Sending message, ID: 11089984-eace-4d06-8a70-f9dbfb0c76af, Expiration: 01:01:45 PST 2016
[00:01:47.136] Close sender connection
[00:01:47.136] Creating sender client
[00:01:47.137] Open sender connection
[00:01:47.137] Sending message, ID: 02dc4930-e1d7-40ad-877c-ef3824a60572, Expiration: 01:01:47 PST 2016
[00:01:48.479] Close sender connection
[00:01:48.480] Creating sender client
[00:01:48.480] Open sender connection
[00:01:48.481] Sending message, ID: aca2eac9-f38b-4c79-8855-7e2caf06f2c7, Expiration: 01:01:48 PST 2016
[00:01:49.496] Close sender connection
[00:01:49.497] Creating sender client
[00:01:49.498] Open sender connection
[00:01:49.498] Sending message, ID: 8ce539b9-6a3b-4639-87fa-7fbb1d14e3ed, Expiration: 01:01:49 PST 2016
[00:01:51.132] Close sender connection
[00:01:51.133] Creating sender client
[00:01:51.133] Open sender connection
[00:01:51.134] Sending message, ID: 82bc2217-af13-45e6-96ce-9e6b11ea206a, Expiration: 01:01:51 PST 2016
[00:01:52.566] Close sender connection
[00:01:52.573] Creating message receiver client
[00:01:52.578] Opening message receiver connection
[00:02:02.575] Creating sender client
[00:02:02.576] Open sender connection
[00:02:02.576] Sending message, ID: 4ef39661-e27c-4074-9fa2-bd9444ebb6da, Expiration: 01:02:02 PST 2016
[00:02:03.498] Message received: Content: test, ID: 4ef39661-e27c-4074-9fa2-bd9444ebb6da
[00:02:03.498] Changing the status of the message to COMPLETE
[00:02:03.574] Close sender connection
[00:02:03.574] Creating sender client
[00:02:03.575] Open sender connection
[00:02:03.575] Sending message, ID: 2d2eba03-8c8d-44da-a2ac-4f25dd49fd8d, Expiration: 01:02:03 PST 2016
[00:02:04.805] Message received: Content: test, ID: 2d2eba03-8c8d-44da-a2ac-4f25dd49fd8d
[00:02:04.806] Changing the status of the message to COMPLETE
[00:02:04.893] Close sender connection
[00:02:08.775] Message received: Content: test, ID: aca2eac9-f38b-4c79-8855-7e2caf06f2c7
[00:02:08.776] Changing the status of the message to COMPLETE
[00:07:04.890] Closing message receiver
[00:07:04.897] Done
```

**Comment by zolvarga** [comment]
Hi Devis,

The sample code you provided create a new Client for every send. Modify the sample to use the existing Client. That would take care about the last three issue you mentioned.

Regarding the first issue: the sender and receiver completely separate objects and we are not trying to synchronize the incoming and outgoing messages.

Thanks,
Zoltan

**Comment by dluc** [comment]
Hi Zoltan,

1.  if the service is still sending messages, these are received before the messages in the queue
2. not all the messages in the queue are processed
3. messages in the queue that happen to be processed are out of order
4. messages are dequeued very slowly

Regarding #1, note that I'm working with a single thread. Are we saying that order is not guaranteed? The [documentation](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messaging) mentions a *queue* so I would assume that messages are delivered in order (e.g. like service bus).

Regarding #2 #3 and #4 could you elaborate please? This is just a sample to reproduce the issue. In reality sender and receiver runs in different locations/processes so I don't understand why using a single sender would solve those issues.

Thanks
Devis

**Comment by dluc** [comment]
After creating a new IoT hub and running the sample code against it, I managed to isolate the root cause. I am generating D2C traffic with a simulator based on [this](https://github.com/Azure/azure-iot-sdk-node/blob/master/device/samples/simple_sample_device.js) code, and after reviewing it I noticed that it was not just sending D2C but also receiving C2D messages, leading to locks and the behavior describe above. Sorry about the oversight.

Thanks
Devis
